src/run_program.rs

```diff
@@ -20,11 +20,39 @@ const OP_COST: Cost = 1;
 const STACK_SIZE_LIMIT: usize = 20000000;
 
 #[cfg(feature = "pre-eval")]
-pub type PreEval =
-    Box<dyn Fn(&mut Allocator, NodePtr, NodePtr) -> Result<Option<Box<PostEval>>, EvalErr>>;
+type PreEvalIndex = usize;
 
 #[cfg(feature = "pre-eval")]
-pub type PostEval = dyn Fn(Option<NodePtr>);
+/// Tell whether to call the post eval function or not, giving a reference id
+/// for the computation to pick up.
+pub enum PreEvalResult {
+    CallPostEval(PreEvalIndex),
+    Done,
+}
+
+#[cfg(feature = "pre-eval")]
+/// Implementing this trait allows an object to be notified of clvm operations
+/// being performed as they happen.
+pub trait PreEval {
+    /// pre_eval is called before the operator is run, giving sexp (the operation
+    /// to run) and args (the environment).
+    fn pre_eval(
+        &mut self,
+        allocator: &mut Allocator,
+        sexp: NodePtr,
+        args: NodePtr,
+    ) -> Result<PreEvalResult, EvalErr>;
+    /// post_eval is called after the operation was performed.  When the clvm
+    /// operation resulted in an error, result is None.
+    fn post_eval(
+        &mut self,
+        _allocator: &mut Allocator,
+        _pass: PreEvalIndex,
```

Comment:
Is this just a nonce to match with the `PreEval` `sexp/args` pair? I suppose `_pass` means something when this is used with the classic compiler, but I wonder if a more generic name might be better.

Reply:
Hmm ya, the name could improve.  I was thinking 'passthrough' but will give it some thought.